File: TypeScript/20.valid-parentheses.ts
/*
 * @lc app=leetcode id=20 lang=typescript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
function isValid(s: string): boolean {
    if (s.length % 2 !== 0) return false

    const map = new Map<string, string>([
        ['(', ')'],
        ['[', ']'],
        ['{', '}']
    ])

    const stack: string[] = []

    for (const char of s){
        if (map.has(char)) stack.push(map.get(char))   
        else if (stack.pop() !== char) return false
    }

    return stack.length === 0
};
// @lc code=end

File: Python/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        d = {
            '(':')',
            '[':']',
            '{':'}'
        }
        temp = []
        for char in s:
            if char in d:
                temp.append(char)
            elif len(temp) >= 1 and char == d[temp.pop()]:
                continue
            else:
                return 0
        if len(temp) == 0:
            return 1
# @lc code=end

